google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/BuildDeploymentSourceType.java

```diff
@@ -0,0 +1,124 @@
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.intellij.appengine.cloud;
+
+import com.intellij.execution.BeforeRunTask;
+import com.intellij.execution.RunManagerEx;
+import com.intellij.execution.configurations.RunConfiguration;
+import com.intellij.execution.impl.ConfigurationSettingsEditorWrapper;
+import com.intellij.ide.DataManager;
+import com.intellij.openapi.actionSystem.DataContext;
+import com.intellij.openapi.module.Module;
+import com.intellij.openapi.project.Project;
+import com.intellij.remoteServer.configuration.deployment.ModuleDeploymentSource;
+import com.intellij.remoteServer.impl.configuration.deployment.DeployToServerRunConfiguration;
+import com.intellij.remoteServer.impl.configuration.deployment.ModuleDeploymentSourceType;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Collection;
+import java.util.List;
+
+import javax.swing.JComponent;
+
+/**
+ * Base class for build system (e.g. maven or gradle) deployment source types.
+ */
+public abstract class BuildDeploymentSourceType extends ModuleDeploymentSourceType {
+
+  @Override
+  public void setBuildBeforeRunTask(
+      @NotNull RunConfiguration configuration,
+      @NotNull ModuleDeploymentSource source) {
+
+    Module module = source.getModule();
+
+    if(module == null) {
+      return;
+    }
+
+    setConfiguration(configuration);
+
+    RunManagerEx runManager = RunManagerEx.getInstanceEx(configuration.getProject());
+    final List<? extends BeforeRunTask> buildTasks = getBuildTasks(runManager, configuration);
+
+    if (hasBuildTaskForModule(buildTasks, module)) {
+      for (BeforeRunTask task : buildTasks) {
+        task.setEnabled(true);
+      }
+    } else {
+      BeforeRunTask buildTask = createBuildTask(module);
+      List<BeforeRunTask> tasks = runManager.getBeforeRunTasks(configuration);
+      tasks.add(buildTask);
+      runManager.setBeforeRunTasks(configuration, tasks, true);
+    }
+  }
+
+  @Override
+  public void updateBuildBeforeRunOption(
+      @NotNull JComponent runConfigurationEditorComponent,
+      @NotNull Project project,
+      @NotNull ModuleDeploymentSource source,
+      boolean select) {
+
+    final DataContext dataContext =
+        DataManager.getInstance().getDataContext(runConfigurationEditorComponent);
+    final ConfigurationSettingsEditorWrapper editor =
+        ConfigurationSettingsEditorWrapper.CONFIGURATION_EDITOR_KEY.getData(dataContext);
+
+    Module module = source.getModule();
+
+    if(module != null && editor != null) {
+      BeforeRunTask buildTask = createBuildTask(module);
+
+      if(buildTask != null) {
+        List<BeforeRunTask> beforeRunTasks = editor.getStepsBeforeLaunch();
+        if (select && !hasBuildTaskForModule(beforeRunTasks, module)) {
+          editor.addBeforeLaunchStep(buildTask);
+        }
+      }
+    }
+  }
+
+  @NotNull
+  protected abstract List<? extends BeforeRunTask> getBuildTasks(
```

Comment:
Yes I agree. The DeploymentSourceType api we have to work with leads to some less than ideal code. I am working on refactoring this and adding documentation.